pkg/controller/helm/helm.go

```diff
@@ -0,0 +1,270 @@
+	"github.com/crossplaneio/gitlab-controller/pkg/meta"
+)
+
+const (
+	controllerName   = "helm." + v1alpha1.APIVersion
+	finalizerName    = "finalizer." + controllerName
+	reconcileTimeout = 5 * time.Minute
+
+	gitlabChartURL = "https://gitlab-charts.s3.amazonaws.com/gitlab-1.8.4.tgz"
+)
+
+// Failure reasons
+const (
+	reasonCreatingResourceProducer    = "cannot create resource producer"
+	reasonProducingResources          = "cannot produce resources"
+	reasonCreatingApplicationProducer = "cannot create application producer"
+	reasonProducingApplication        = "cannot produce application"
+	reasonSyncingApplication          = "cannot sync application"
+)
+
+var (
+	log = logging.Logger.WithName(controllerName)
+)
+
+// Suggested reconcile results. TODO(negz): Dedupe if/when
+// https://github.com/crossplaneio/crossplane/pull/424/ is merged.
+var (
+	RequeueIfDirty = reconcile.Result{Requeue: false}
+	RequeueNow     = reconcile.Result{Requeue: true}
+	RequeueSoon    = reconcile.Result{RequeueAfter: 30 * time.Second}
+	RequeueLater   = reconcile.Result{RequeueAfter: 2 * time.Minute}
+)
+
+// Add creates a new GitLab Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+
+	// TODO(negz): Should we add these types to our manager here (close to where
+	// they're used, at risk of duplicates) or all together somewhere?
+	if err := xpworkloadv1alpha1.SchemeBuilder.AddToScheme(mgr.GetScheme()); err != nil {
+		return errors.Wrapf(err, "cannot add %s to scheme", xpworkloadv1alpha1.APIVersion)
+	}
+
+	r := &Reconciler{
+		kube:  combined.NewClient(mgr.GetClient()),
+		local: &localCluster{kube: mgr.GetClient()},
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return errors.Wrap(err, "cannot create controller")
+	}
+
+	return errors.Wrapf(c.Watch(&source.Kind{Type: &v1alpha1.GitLab{}}, &handler.EnqueueRequestForObject{}), "cannot watch for %s", v1alpha1.GitLabKind)
+}
+
+// A ResourceProducer produces Kubernetes resources.
+type ResourceProducer interface {
+	// Produce Kubernetes resources as a slice of *unstructured.Unstructured.
+	Produce() ([]*unstructured.Unstructured, error)
+}
+
+// ApplicationProducer produces KubernetesApplications.
+type ApplicationProducer interface {
+	// Produce a KubernetesApplication.
+	Produce() (*xpworkloadv1alpha1.KubernetesApplication, error)
+}
+
+type syncer interface {
+	sync(ctx context.Context, gl *v1alpha1.GitLab) reconcile.Result
+}
+
+type deleter interface {
+	delete(ctx context.Context, gl *v1alpha1.GitLab) reconcile.Result
+}
+
+type syncDeleter interface {
+	syncer
+	deleter
+}
+
+type localCluster struct {
+	kube client.Client
+}
+
+func (c *localCluster) sync(ctx context.Context, gl *v1alpha1.GitLab) reconcile.Result {
+	meta.AddFinalizer(gl, finalizerName)
```

Comment:
Why do we need finalizer in this case?
For managed resources (in crossplane), we need it to handle resource deletion.
Here, since we are using `OwnerReference` all "childre" object should be deleted and garbage collected via "backgroud delete".

Reply:
Force of habit. This will go away since we're merging into your controller.